lib/helpers/signed-aws-es-fetch.js

```diff
@@ -0,0 +1,109 @@
+const aws4 = require('aws4');
+const nodeFetch = require('node-fetch');
+const logger = require('@financial-times/n-logger').default;
+const resolveCname = require('util').promisify(require('dns').resolveCname);
+const memoize = require('lodash/memoize');
+
+const memoizedResolveCname = memoize(resolveCname);
+
+const envKeys = {
+	awsKeys: ['ES_AWS_ACCESS_KEY', 'AWS_ACCESS_KEY', 'AWS_ACCESS_KEY_ID'],
+	awsSecretKeys: ['ES_AWS_SECRET_ACCESS_KEY', 'AWS_SECRET_ACCESS_KEY'],
+	tokenKeys: ['ES_AWS_SESSION_TOKEN', 'AWS_SESSION_TOKEN']
+};
+
+module.exports = async function (url, opts, creds) {
+	opts = opts || {};
+	url = await resolveUrlAndHost(url);
+	return signedFetch(url, opts, creds);
+};
+
+function getURL (url) {
+	try {
+		const urlObject = new URL(url);
+		return urlObject;
+	} catch (error) {
+		logger.error({
+			event: 'Invalid URL',
+			data: {
+				url
+			}
+		});
+		throw error;
+	}
+}
+
+async function resolveValidHost (urlObject) {
+	const hosts = await memoizedResolveCname(urlObject.host);
+	if (hosts === undefined) {
+		logger.error({
+			event: 'Invalid Host',
+			data: {
+				url: urlObject
+			}
+		});
+		throw new Error('Invalid Host');
+	}
+	return hosts[0];
+}
+
+async function resolveUrlAndHost (url) {
+	const urlObject = getURL(url);
+	if (
+		!/\.es\.amazonaws\.com$/.test(urlObject.host) &&
+		!process.env.AWS_SIGNED_FETCH_DISABLE_DNS_RESOLUTION
+	) {
+		const host = await resolveValidHost(urlObject);
+		url = url.replace(urlObject.host, host);
+	}
+	return url;
+}
+
+function getEnv (keys) {
+	const keyName = envKeys[keys].find((keyName) => process.env[keyName]);
+	return process.env[keyName];
+}
+
+function setAwsCredentials (creds) {
+	creds = creds || {};
+	creds.accessKeyId = creds.accessKeyId || getEnv('awsKeys');
+	creds.secretAccessKey = creds.secretAccessKey || getEnv('awsSecretKeys');
+
+	let sessionToken = creds.sessionToken;
+	if (
+		process.env.ES_AWS_SESSION_TOKEN !== false &&
+		// a boolean value is interpreted as string if set from vault
+		process.env.ES_AWS_SESSION_TOKEN !== 'false'
+	) {
+		sessionToken = sessionToken || getEnv('tokenKeys');
+	}
+	if (sessionToken) {
+		creds.sessionToken = sessionToken;
+	}
+	return creds;
+}
+
+function getSignableData (url, opts, creds) {
+	creds = setAwsCredentials(creds);
+	const { host, pathname: path, search, protocol } = new URL(url);
```

Comment:
nitpick:  In the end, the URL object creates the host property by parsing the url address.
So possible cases:

1. We get an error because the url is invalid.
2. We get an error because we don't resolve a valid canonical host (resolveCname)
3. We use a valid url and a valid canonical host url. 
Host property format is valid and equal for both so there is no way to get an error in url.replace(urlObject.host, host) unless some extremely rare DNS  corruption/hacking case for which is going to fail also the fetch operation.

Anyway, I agree with you. I think it's better for consistency to use getUrl.